services/graph/pkg/service/v0/sharedwithme.go

```diff
@@ -0,0 +1,235 @@
+package svc
+
+import (
+	"context"
+	"net/http"
+	"strings"
+
+	userpb "github.com/cs3org/go-cs3apis/cs3/identity/user/v1beta1"
+	rpc "github.com/cs3org/go-cs3apis/cs3/rpc/v1beta1"
+	collaboration "github.com/cs3org/go-cs3apis/cs3/sharing/collaboration/v1beta1"
+	storageprovider "github.com/cs3org/go-cs3apis/cs3/storage/provider/v1beta1"
+	"github.com/cs3org/reva/v2/pkg/storagespace"
+	"github.com/go-chi/render"
+	libregraph "github.com/owncloud/libre-graph-api-go"
+
+	"github.com/owncloud/ocis/v2/services/graph/pkg/identity"
+	"github.com/owncloud/ocis/v2/services/graph/pkg/service/v0/errorcode"
+)
+
+// ListSharedWithMe lists the files shared with the current user.
+func (g Graph) ListSharedWithMe(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	driveItems, err := g.listSharedWithMe(ctx)
+	if err != nil {
+		g.logger.Error().Err(err).Msg("listSharedWithMe failed")
+		errorcode.RenderError(w, r, err)
+		return
+	}
+
+	render.Status(r, http.StatusOK)
+	render.JSON(w, r, &ListResponse{Value: driveItems})
+}
+func (g Graph) listSharedWithMe(ctx context.Context) ([]libregraph.DriveItem, error) {
+	gatewayClient, err := g.gatewaySelector.Next()
+	if err != nil {
+		g.logger.Error().Err(err).Msg("could not select next gateway client")
+		return nil, err
+	}
+
+	listReceivedSharesResponse, err := gatewayClient.ListReceivedShares(ctx, &collaboration.ListReceivedSharesRequest{})
+	if err != nil {
+		g.logger.Error().Err(err).Msg("listing shares failed")
+		return nil, errorcode.New(errorcode.GeneralException, err.Error())
+	}
+
+	switch listReceivedSharesResponse.Status.Code {
+	case rpc.Code_CODE_NOT_FOUND:
+		return nil, identity.ErrNotFound
+	}
+
+	var driveItems []libregraph.DriveItem
+	for _, receivedShare := range listReceivedSharesResponse.GetShares() {
+		share := receivedShare.GetShare()
+		if share == nil {
+			continue
+		}
+
+		driveItem := &libregraph.DriveItem{}
+
+		statResponse, err := gatewayClient.Stat(ctx, &storageprovider.StatRequest{Ref: &storageprovider.Reference{ResourceId: share.GetResourceId()}})
+		if err != nil {
+			g.logger.Error().Err(err).Msg("could not stat")
+			continue
+		}
+		if statResponse.GetStatus().GetCode() != rpc.Code_CODE_OK {
+			g.logger.Error().Err(err).Msg("invalid stat response")
+			continue
+		}
+		resourceInfo := statResponse.GetInfo()
+
+		var driveOwner *libregraph.Identity
+		if user := statResponse.GetInfo().GetOwner(); user != nil {
+			if getUserResponse, err := gatewayClient.GetUser(ctx, &userpb.GetUserRequest{
```

Comment:
You might want to switch this to `g.identityCache.GetUser(ctx, user.GetOpaqueId())`, which provides some simple caching.

If not you'd need to at least check `getUserResponse.Status.Code` if the request actually succeed. `err` just indicates lower level errors (e.g. network issue and related stuff) AFAIK.

Reply:
done